src/client/inference_session.py

```diff
@@ -154,60 +157,145 @@ def __exit__(self, *exc_details):
+                    rpc_info=self._sequence_manager.rpc_info,
+                    timeout=self._sequence_manager.timeout,
+                    max_length=self._max_length,
+                    **self._metadata,
                 )
             )
-            self.inference_sessions.append(inference_session)
-            self.stack.enter_context(inference_session)
+            server_sessions.append(session)
+            session.__enter__()
+        return server_sessions
 
+    def _exit_server_sessions(self, server_sessions: List[_ServerInferenceSession]) -> None:
+        for session in reversed(server_sessions):
+            try:
+                session.__exit__(None, None, None)
+            except Exception:
+                logger.debug("Caught exception while closing connection to server:", exc_info=True)
+
+    def __enter__(self):
+        assert not self._closed and not self._chosen_spans
         return self
 
     def step(self, inputs: torch.Tensor, prompts: Optional[torch.Tensor] = None, **kwargs):
-        assert not self.closed
+        assert not self._closed
         if torch.is_grad_enabled():
             logger.warning("Running inference session with grad enabled. Gradients will *not* be propagated correctly.")
         if prompts is None or is_dummy(prompts):
             prompts = DUMMY
         else:
-            assert prompts.ndim == 4 and prompts.shape[0] == len(self.sequence_manager)
-        for session in self.inference_sessions:
-            outputs = session.step(inputs, prompts[self.chosen_spans[0].start : self.chosen_spans[0].end], **kwargs)
-            assert outputs.shape == inputs.shape, f"expected {inputs.shape}, got {outputs.shape}"
-            inputs = outputs
+            assert prompts.ndim == 4 and prompts.shape[0] == len(self._sequence_manager)
+        n_input_tokens = inputs.shape[1]
+        if self._position + n_input_tokens > self._max_length:
+            raise ValueError(
+                f"Maximum length exceeded: prefix {self._position} + current {n_input_tokens} exceeds pre-allocated maximum {self._max_length}"
+            )
+
+        server_idx = 0
+        block_idx = 0
+        recovery_end = None  # Recovery mode is disabled until a failure happens
+        while block_idx < len(self._sequence_manager):
+            for attempt_no in itertools.count():
+                logger.debug(f"Inference: block {block_idx}, attempt {attempt_no}")
+                try:
+                    if attempt_no >= 1:
+                        self._sequence_manager.update_()
+                        if server_idx < len(self._chosen_spans):
+                            recovery_end = self._chosen_spans[server_idx].end
+                        else:
+                            recovery_end = len(self._sequence_manager)
+                        if attempt_no == 1:
+                            logger.info(
+                                f"Entering recovery mode, remote attention caches "
```

Comment:
[nit] "entering recovery mode" reads as "from now on, all subsequent actions will happen in this new mode". Maybe "One of remote servers failed, recovering..." ?